pm: Make device pm optional per state

ceolin opened the pull request:
This is a proposal to make device power management (triggered when SoC is idle) completely optional.

- Targets can select which power states should trigger device power management
- If there is no power state with this property enabled, the pm subsystem won't trigger device power management
- Removed the need for `CONFIG_PM_DEVICE_RUNTIME_EXCLUSIVE`

Notes:

This draft contains a proposal but it is not complete, ~~some tests are failing because of the latest commit that completely disable device pm when it is not enabled in DT and we have some tests that don't use DT to check pm API.~~ This is commit is basically an additional optimization to have the same we had with `CONFIG_PM_DEVICE_RUNTIME_EXCLUSIVE`.

~~In case of consensus, I will restore the current behavior adding `zephyr,pm-device-enabled` in all power states (across the tree) that are currently triggering device power management, then we can in a case by case disable it if need. But the idea is not breaking the current behavior.~~

Additional context:
- https://github.com/zephyrproject-rtos/zephyr/pull/60463
- https://github.com/zephyrproject-rtos/zephyr/pull/60369

Commit:
pm: Remove PM_DEVICE_RUNTIME_EXCLUSIVE

That is option has shown confusing on it is attempt to prevent
system pm doing device power management. Lets address this
problem properly.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
pm: Enable device pm per power state

Enable device power management individually per power state.
This allows targets tuning which state should trigger
device power management or completely disable it simply not
enabling it in any power state.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
pm: Exclude device pm path when it is not needed

Remove device pm path when there is no is no power state in DT with
device pm enabled. This basically does the same thing that was done
by PM_DEVICE_RUNTIME_EXCLUSIVE.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

JordanYates commented on include/zephyr/pm/state.h:
```diff
@@ -154,6 +155,15 @@ struct pm_state_info {
+	 * will suspend devices before entering this state and will
+	 * properly resume them when leaving it.
+	 */
+	bool pm_device_enabled;
```
Potentially memory usage could be optimised by placing this with `uint8_t substate_id;` to reduce padding

ceolin replied:
yep, that makes sense, this way  we just use the space in the current padding,

Commit:
tests: power_mgmt: Use DTS overlay for power state

Define the power state needed in the test in DT.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
tests: device_wakeup_api: Use DTS overlay for power state

Define the power state needed in the test in DT.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
doc: pm: Update system device pm

Add new information about new property to
enable/disable device power management per
power state.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

knthm commented:
I think this is a very nice non-destructive compromise of #60463.

The ability to individually disable device PM for peripherals where it might conflict with system PM is a great idea! :)

ceolin commented:
> I think this is a very nice non-destructive compromise of #60463.
> 
> The ability to individually disable device PM for peripherals where it might conflict with system PM is a great idea! :)

Thanks for looking it.

gmarull reviewed (changes requested):
I think this PR doesn't solve the problems in device PM itself. At least with `CONFIG_PM_DEVICE_RUNTIME_EXCLUSIVE=y` you had the guarantee that `CONFIG_PM_DEVICE` did nothing, now, as soon as you have a single state with `zephyr,pm-device-enabled` and a device with a blocking PM callback you're screwed again.

The main problem with `CONFIG_PM_DEVICE` is not that it doesn't allow blocking, but that it treats all devices equally. If we really want to keep this mode, I think it needs its own set of PM callbacks (where you can ask clients to not block), separate from PM device runtime (as on Linux, actually). But on MCUs this PM model doesn't make sense, you typically don't want to deal with devices on every LP state transition. Some MCUs need to perform actions when exiting certain LP states on a per-device basis (e.g. the STOP modes in STM32), but this is something that can be solved in a better way, for example, subscribing to such events as needed. And to me, such a mechanism could likely be re-used to finally get rid of CONFIG_PM_DEVICE. The problem here is that we've never got technical details on how it is being used to discuss a proper way forward, and we've been stuck forever with this.

ceolin commented:
> I think this PR doesn't solve the problems in device PM itself. At least with `CONFIG_PM_DEVICE_RUNTIME_EXCLUSIVE=y` you had the guarantee that `CONFIG_PM_DEVICE` did nothing, now, as soon as you have a single state with `zephyr,pm-device-enabled` and a device with a blocking PM callback you're screwed again.

That is not what this pr aims. The goal here is bring more flexibility for those who wants to use it and not breaking the current behavior. I can restore `PM_DEVICE_RUNTIME_EXCLUSIVE` though, since this pr does not easily replace it indeed.

> 
> The main problem with `CONFIG_PM_DEVICE` is not that it doesn't allow blocking, but that it treats all devices equally. If we really want to keep this mode, I think it needs its own set of PM callbacks (where you can ask clients to not block), separate from PM device runtime (as on Linux, actually). But on MCUs this PM model doesn't make sense, you typically don't want to deal with devices on every LP state transition. 

That is one reason to allow them to choose only low power states that it should be triggered.

>Some MCUs need to perform actions when exiting certain LP states on a per-device basis (e.g. the STOP modes in STM32), but this is something that can be solved in a better way, for example, subscribing to such events as needed. And to me, such a mechanism could likely be re-used to finally get rid of CONFIG_PM_DEVICE. 

notifiers don't solve it now since they can't block as well (they are called from idle thread)


> The problem here is that we've never got technical details on how it is being used to discuss a proper way forward, and we've been stuck forever with this.

We can work through this, but there are real use cases. Note that this is bring more flexibility making even easier to not  use it if that is wanted. We may work through a better way but this change is just giving the capability to tell whether or not a state triggers device pm.

Is removing `CONFIG_PM_DEVICE_RUNTIME_EXCLUSIVE` the reason for blocking it ? If that is the case it can easily be restored. Other than that this is not doing anything other than make easier to avoid system device pm since one can now select only states (or none) that this should happen.

ceolin commented:
@gmarull ping

knthm reviewed (changes requested):
Inverting the property name and behavior could be lower impact, i.e. `zephyr,pm-device-disabled`.
That'd also guarantee identical behavior to before without SoC DTS changes.

The use cases in which a `power-state` would require disabling device pm seems far fewer than those that need/want it enabled.

knthm commented:
> I think this PR doesn't solve the problems in device PM itself. At least with CONFIG_PM_DEVICE_RUNTIME_EXCLUSIVE=y you had the guarantee that CONFIG_PM_DEVICE did nothing, now, as soon as you have a single state with zephyr,pm-device-enabled and a device with a blocking PM callback you're screwed again.

> The main problem with CONFIG_PM_DEVICE is not that it doesn't allow blocking, but that it treats all devices equally. If we really want to keep this mode, I think it needs its own set of PM callbacks (where you can ask clients to not block), separate from PM device runtime (as on Linux, actually).

> notifiers don't solve it now since they can't block as well (they are called from idle thread)

A huge issue with device-specific PM behavior, whether reinit or suspend/resume, is that they run in idle/handler mode threads/ISRs, which can't block.

For the reinit stuff I've run into various race conditions due to this, as the logic there generally needs to block while waiting for shared hardware resources. This applies to PM Device actions too. IMHO, we need to provide any hardware-specific logic with the ability to block.

My solution, which I'll be pushing after some cleanup, is to delegate the reinit logic to the system workqueue.
I wonder if it might be a good idea to add a special non-yielding/high-priority PM-specific workqueue. Another option might be to provide a special flag that disables yielding for a specific system workqueue item.

The only other alternative I can see, would be adding some more complex signaling logic into the kernel for the main thread that checks if hardware-specific actions are required before resuming normal application execution (I'm not familiar enough with the task switching and kernel main thread behavior to suggest where this could go, or how (im)possible this might be).

> But on MCUs this PM model doesn't make sense, you typically don't want to deal with devices on every LP state transition. Some MCUs need to perform actions when exiting certain LP states on a per-device basis (e.g. the STOP modes in STM32)

You kind of do, at least sometimes. Select peripherals on STM32 are still active in various STOP-modes, if you want to use this peripheral only during run-mode, you'll want to trigger device suspend/resume during PM transitions to save some additional power.

___

Personally, this PR is still a very nice stopgap between the current on/off `CONFIG_PM_DEVICE_RUNTIME_EXCLUSIVE` behavior which is buggy, or even completely unusable for some power states (STOP2 on STM32WL), and removing it entirely, which breaks many current uses of it, like SoF.

knthm commented on include/zephyr/pm/state.h:
```diff
@@ -209,6 +219,7 @@ struct pm_state_info {
 		.substate_id = DT_PROP_OR(node_id, substate_id, 0),	       \
 		.min_residency_us = DT_PROP_OR(node_id, min_residency_us, 0),  \
 		.exit_latency_us = DT_PROP_OR(node_id, exit_latency_us, 0),    \
+		.pm_device_enabled = DT_PROP_OR(node_id, zephyr_pm_device_enabled, false),    \
```
Line break after `node_id,` for escape-character line-length consistency?

ceolin commented:
> > > I think this PR doesn't solve the problems in device PM itself. At least with CONFIG_PM_DEVICE_RUNTIME_EXCLUSIVE=y you had the guarantee that CONFIG_PM_DEVICE did nothing, now, as soon as you have a single state with zephyr,pm-device-enabled and a device with a blocking PM callback you're screwed again.
> > 
> > 
> > > The main problem with CONFIG_PM_DEVICE is not that it doesn't allow blocking, but that it treats all devices equally. If we really want to keep this mode, I think it needs its own set of PM callbacks (where you can ask clients to not block), separate from PM device runtime (as on Linux, actually).
> > 
> > 
> > > notifiers don't solve it now since they can't block as well (they are called from idle thread)
> 
> that's true, they simply decouple system PM from devices.
> 
> > A huge issue with device-specific PM behavior, whether reinit or suspend/resume, is that they run in idle/handler mode threads/ISRs, which can't block.
> > For the reinit stuff I've run into various race conditions due to this, as the logic there generally needs to block while waiting for shared hardware resources. This applies to PM Device actions too. IMHO, we need to provide any hardware-specific logic with the ability to block.
> 
> It'd be nice to have some concrete examples.
> 
> > My solution, which I'll be pushing after some cleanup, is to delegate the reinit logic to the system workqueue. I wonder if it might be a good idea to add a special non-yielding/high-priority PM-specific workqueue. Another option might be to provide a special flag that disables yielding for a specific system workqueue item.
> 
> I guess that would need to come with one restriction: you can't go back to _regular threads_ until finished; otherwise, you'd need to start guarding many drivers with a kind of "wait_ready()" because you could have user code ready for execution before driver re-init finishes.
> 
> > The only other alternative I can see, would be adding some more complex signaling logic into the kernel for the main thread that checks if hardware-specific actions are required before resuming normal application execution (I'm not familiar enough with the task switching and kernel main thread behavior to suggest where this could go, or how (im)possible this might be).
> > > But on MCUs this PM model doesn't make sense, you typically don't want to deal with devices on every LP state transition. Some MCUs need to perform actions when exiting certain LP states on a per-device basis (e.g. the STOP modes in STM32)
> > 
> > 
> > You kind of do, at least sometimes. Select peripherals on STM32 are still active in various STOP-modes, if you want to use this peripheral only during run-mode, you'll want to trigger device suspend/resume during PM transitions to save some additional power.
> 
> The whole point of device runtime PM is to have devices enabled only when needed, so at the moment you go to idle, your device will only be ON if it's waiting on a pending transaction; otherwise it will already be in a low-power state. Your device is also responsible for forbidding _dangerous_ states by locking them, e.g. preventing context loss if waiting for a transaction completion. Device-specific actions related to system PM only make sense if there's a relationship between those, e.g. you loose context, some special flag needs to be set, etc. I think notifiers are a better fit for this problem, you just subscribe if you really need it.
> 
> > Personally, this PR is still a very nice stopgap between the current on/off `CONFIG_PM_DEVICE_RUNTIME_EXCLUSIVE` behavior which is buggy, or even completely unusable for some power states (STOP2 on STM32WL), and removing it entirely, which breaks many current uses of it, like SoF.
> 
> To me, this PR is just adding yet another layer of complexity to a model that doesn't work well and can not solve some basic use cases. I fear it will take yet a few more years to realize and move to something better.

That is fine, if we get to this conclusion and we get something better we just move forward. This "additional complexity is almost entirely handled by us in the pm subsystem, and it is a step forward having less use of system device pm. We have same goals, the difference is that I don't want to remove this feature now because there are targets using it yet and because for some devices this simple model can work, if they don't need to do anything that requires blocking during pm action, it is way simpler than supporting device runtime.

Let me add another thing, device runtime pm is not a silver bullet that works for all cases. There are devices that have to be active until the soc suspension (or that are even used during the suspension process), in this scenario even if they use device runtime pm we end up having the same blocking issue.

Assuming that having device runtime pm only will lead to *ALL* devices be suspended when the SoC is a big shoot ... if we get there, we remove this other model, in the meantime what I am proposing with this pr is to use less this system device pm model, targets that need / want use this model have to do it explicitly now.

ceolin commented:
> Inverting the property name and behavior could be lower impact, i.e. `zephyr,pm-device-disabled`. That'd also guarantee identical behavior to before without SoC DTS changes.
> 
> The use cases in which a `power-state` would require disabling device pm seems far fewer than those that need/want it enabled.

That's is currently true, but the idea here is to push toward device runtime pm, still allowing using system device pm. Hopefully in future these will be just few cases, that is why I changed it to make targets need to explicitly enabling it when they need. The fact that I have done it for many targets is to keep the current behavior, but it needs to be reviewed case by case.

knthm commented:
> That's is currently true, but the idea here is to push toward device runtime pm, still allowing using system device pm. Hopefully in future these will be just few cases, that is why I changed it to make targets need to explicitly enabling it when they need. The fact that I have done it for many targets is to keep the current behavior, but it needs to be reviewed case by case.

I see, that makes perfect sense. In that case, disregard my comment on inverting the property behavior.

knthm commented on subsys/pm/pm.c:
```diff
@@ -238,8 +245,8 @@ bool pm_system_suspend(int32_t ticks)
 	/* Wake up sequence starts here */
-#if defined(CONFIG_PM_DEVICE)
 	if (atomic_add(&_cpus_active, 1) == 0) {
 		pm_resume_devices();
```
One small issue I noticed testing this today:

The `pm_resume_devices()` call here may still run when the low power state we're returning from doesn't have the `pm_device_enabled` flag set, and thus `pm_resume_devices()` above wasn't called.
So in DT configurations where not all low power states have said flag, we'll be executing device pm resume logic unnecessarily for devices that hadn't been suspended.

Maybe we could locally retain the `z_cpus_pm_state[id].pm_device_enabled` bool check above and use that here to only resume devices that have actually been suspended by system pm?

knthm replied:
You're right of course.
I should've checked the device resume function more closely.

erwango reviewed (commented):
Non blocking but a couple of comments in the sake of giving some clarity on the feature

ceolin commented on include/zephyr/pm/state.h:
```diff
@@ -134,12 +134,22 @@ struct pm_state_info {
+	 * will suspend devices before entering this state and will
+	 * properly resume them when leaving it.
+	 */
+	bool pm_device_enabled;
```
this is a field of `pm_state_info`, other fields don't have this suffix / prefix. It seems redundant imho, but I can add it if it makes clear.

erwango replied:
Agree this would be redundant, but could add some clarity to users.

ceolin replied:
ok, changed it.

ceolin commented on doc/services/pm/device.rst:
```diff
@@ -44,12 +44,39 @@ When using this type, device power management is mostly done inside
 possible dependencies between them are satisfied. As soon as the CPU wakes up
 from a sleep state, devices are resumed in the opposite order that they were
 suspended.
 
```
Example about this model ? Or about enabling/disabling it per power state ?

If it is about the model, there are some examples in lines 86/88.

erwango replied:
ok, right, this should be sufficient.

ceolin commented on dts/bindings/power/zephyr,power-state.yaml:
```diff
@@ -31,3 +31,9 @@ properties:
+  zephyr,pm-device-enabled:
+    type: boolean
+    description: |
+       Trigger device power management for this state. When this is enabled
+       the power management will suspend devices before entering
+       this state and resume them when leaving it.
```
you are right, I think it makes a lot of sense.

ceolin commented:
@nashif  @erwango can you take a look again please?

gmarull reviewed (dismissed):
Posted an alternative for the exclusive removal https://github.com/zephyrproject-rtos/zephyr/pull/62160, ie, PM device runtime can be enabled standalone, so developers don't need to deal with all the flaws associated with CONFIG_PM_DEVICE.

As I said multiple times, the link between devices and system power states needs to be solved differently, e.g. modifying the notifiers API so that devices that need to take actions when the system enters/exits some lp states can do so. CONFIG_PM_DEVICE is a flawed model that has led to a lot of confusion and pain, extending it will just perpetuate the problem with yet another hack.

knthm commented:
@gmarull in line with that, could you review #60998?

gmarull commented:
> @gmarull in line with that, could you review #60998?

sure, will do

ceolin commented:
> Posted an alternative for the exclusive removal #62160, ie, PM device runtime can be enabled standalone, so developers don't need to deal with all the flaws associated with CONFIG_PM_DEVICE.

Gosh, this is not the same thing. This pr is not doing only it, what is in that pr is not addressing the problem that this pr is doing. @gmarull This pr is letting people that wants to use system device pm define which states should trigger this action, your pr is de-coupling pm device from device runtime pm. These are not the same thing.

If this model is not suitable don't use, it is simple as that, these changes complement each other, but this pr is adding more control to system device pm  and that's all.

> 
> As I said multiple times, the link between devices and system power states needs to be solved differently, e.g. modifying the notifiers API so that devices that need to take actions when the system enters/exits some lp states can do so. CONFIG_PM_DEVICE is a flawed model that has led to a lot of confusion and pain, extending it will just perpetuate the problem with yet another hack.

Again, the main problem is blocking while the system is going to sleep, notifiers don't solve this problem because they are called from the same context. device runtime is comfortably assuming that devices will be already suspended when the system gets idle, but that is not true in some cases, there are cases where the device has to be active waiting external event until the system goes to sleep.

Commit:
pm: device: De-couple device pm from system pm

PM_DEVICE is not attached to system managed device power management.
It is a very common use case targets with device runtime power
management that don't want system device power management enabled.

We introduce a new symbol (PM_DEVICE_SYSTEM_MANAGED) to explicit
control whether or not system device power management should be
globally enabled.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

ceolin commented:
@nashif @JordanYates @erwango Would you mind taking a look on it again ? I have added one important change here that is de-coupling device power management from system device power management. In general it achieves something similar to `PM_DEVICE_RUNTIME_EXCLUSIVE` but is more clear and separates the support for device power management from the feature  that allows the system suspend / resume devices.

In resume with this this pr we can:

- Select which power states should trigger system device power management
- Globally disable this behavior (without need a dt overlay)

Added DNM label to avoid it be merged without more eyes. Will add explicit test for this new option as well.

Commit:
tests: power_mgmt: Add test for PM_DEVICE_SYSTEM_MANAGED

Add a new test that disables PM_DEVICE_SYSTEM_MANAGED, since this
option is enabled by default.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

ceolin commented:
@nashif @erwango @JordanYates please take a look, this is introducing some core changes.

Commit:
west: sof: Update SoF

Bump SoF to a version that does not have reference to
PM_DEVICE_RUNTIME_EXCLUSIVE

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

erwango commented on subsys/pm/Kconfig:
```diff
@@ -114,6 +113,16 @@ config PM_DEVICE_RUNTIME
+	help
+	  This option enables the device system managed power management.  The
+	  power management subsystem will suspend devices when it gets idle
+	  and the SoC power state set has power management enabled. In opposite,
```
```
	  and if the SoC power state set has power management enabled. In opposite,
```

erwango commented on doc/services/pm/device.rst:
```diff
@@ -37,29 +37,67 @@ For more information, see :ref:`pm-device-runtime`.
+   In this scenario, some devices have to be always active and should be suspended together with
+   the SoC when requested by this external entity.
+
+It is important to emphasize that his method has, already mentioned,
```
```suggestion
It is important to emphasize that this method has, as already mentioned,
```

JordanYates commented on subsys/pm/device_system_managed.c:
```diff
@@ -0,0 +1,90 @@
+			return false;
+		}
+
+		TYPE_SECTION_START(pm_device_slots)[num_susp] = dev;
```
Why is this updating an array in RAM instead of setting a flag on the device PM struct?

ceolin replied:
The idea was to have a list of only suspended devices to resume (in reverse order) when the system is resumed avoiding need to check all devices.
Honestly, thinking now it seems that paying this price is better than having this pre-allocated array in ram ...

This logic was not introduced in these changes though, we can review it separately I believe.

ceolin commented on include/zephyr/pm/state.h:
```diff
@@ -209,6 +219,7 @@ struct pm_state_info {
 		.substate_id = DT_PROP_OR(node_id, substate_id, 0),	       \
 		.min_residency_us = DT_PROP_OR(node_id, min_residency_us, 0),  \
 		.exit_latency_us = DT_PROP_OR(node_id, exit_latency_us, 0),    \
+		.pm_device_enabled = DT_PROP_OR(node_id, zephyr_pm_device_enabled, false),    \
```
sure thing

erwango commented on samples/boards/stm32/power_mgmt/blinky/prj.conf:
```diff
@@ -1,5 +1,4 @@
 CONFIG_PM=y
 CONFIG_PM_DEVICE=y
 CONFIG_PM_DEVICE_RUNTIME=y
-CONFIG_PM_DEVICE_RUNTIME_EXCLUSIVE=n
```
To be replaced by `PM_DEVICE_SYSTEM_MANAGED=y` then ?

ceolin replied:
that is enabled by default.

erwango commented on samples/boards/stm32/power_mgmt/suspend_to_ram/prj.conf:
```diff
@@ -1,7 +1,6 @@
 CONFIG_PM=y
 CONFIG_PM_DEVICE=y
 CONFIG_PM_DEVICE_RUNTIME=y
-CONFIG_PM_DEVICE_RUNTIME_EXCLUSIVE=n
```
Ditto

ceolin replied:
same

JordanYates commented on subsys/pm/Kconfig:
```diff
@@ -114,13 +113,15 @@ config PM_DEVICE_RUNTIME
-config PM_DEVICE_RUNTIME_EXCLUSIVE
-	depends on PM_DEVICE_RUNTIME
-	bool "Use only on Runtime Power Management on system suspend / resume"
+config PM_DEVICE_SYSTEM_MANAGED
```
Isn't this changing the default behaviour when `PM_DEVICE=y` and `PM_DEVICE_RUNTIME=y`?
Previously hooks wouldn't be enabled, now they are.

ceolin replied:
`PM_DEVICE_RUNTIME_EXCLUSIVE=y` can be replace with `PM_DEVICE_SYSTEM_MANAGED=n`.

I believe this is more clear. 

`PM_DEVICE` is the base and need by RUNTIME and SYSTEM_MANAGED. `PM_DEVICE` is the base to enable pm in a device (aka pm callback)

`PM_DEVICE_RUNTIME` enables device runtime
`PM_DEVICE_SYSTEM_MANAGED` tell the pm subsystem to suspend / resume devices when the soc goes idle.

If one does not want device system managed pm, just disable it. We can make `default n if PM_DEVICE_RIUNTIME`

JordanYates replied:
I understand it can be replaced, but `PM_DEVICE_RUNTIME_EXCLUSIVE=y` was  the previous default, and `PM_DEVICE_SYSTEM_MANAGED=y` is the new default, which is the opposite behaviour.

ceolin replied:
Right ... but now we have to explicitly set, in DT, which states triggers system managed device pm.

If we want to keep exactly the same behavior, I can change this pr to;

1) By default, states trigger device pm when `PM_DEVICE_SYSTEM_MANAGED` is set (and users can disable specific sates setting it in DT)
2) PM_DEVICE_SYSTEM_MANAGED is disabled (by default) when PM_DEVICE_RUNTIME is enabled.

With these changes we keep the current behavior. I was already thinking about it (especially due dt changes).
@JordanYates what you think ?

henrikbrixandersen commented:
@ceolin I see you have added this to the v3.6.0 milestone. Do you still intend for this to go in before v3.6.0? It looks like an enhancement to me.

ceolin commented:
> @ceolin I see you have added this to the v3.6.0 milestone. Do you still intend for this to go in before v3.6.0? It looks like an enhancement to me.

@henrikbrixandersen it won't. It is to late for it now. I'll wait the merge window open to move it.

ceolin commented:
@JordanYates I have kept the current behavior in https://github.com/zephyrproject-rtos/zephyr/pull/70623. Please take a look.

ceolin commented:
Closing it in favor of #70623
